x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java

```diff
@@ -1183,4 +1205,207 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+                    assertEquals(asyncStatus.get(IS_RUNNING_NAME), asyncStatus.get(IS_PARTIAL_NAME));
+                    assertEquals(asyncId, asyncStatus.get(ID_NAME));
+                }
+
+                // fetch the results (potentially waiting now to complete)
+                Request request = new Request("GET", SQL_ASYNC_REST_ENDPOINT + asyncId);
+                if (pollForCompletion == false) {
+                    request.addParameter(WAIT_FOR_COMPLETION_TIMEOUT_NAME, "1d");
+                }
+                if (randomBoolean()) {
+                    request.addParameter(URL_PARAM_FORMAT, format);
+                    if (format.equals("csv")) {
+                        csvDelimiter = ';';
+                        request.addParameter(URL_PARAM_DELIMITER, URLEncoder.encode(String.valueOf(csvDelimiter), StandardCharsets.UTF_8));
+                    }
+                } else {
+                    request.setOptions(request.getOptions().toBuilder().addHeader("Accept", acceptMap.get(format)));
+                }
+                response = client().performRequest(request);
+
+                assertEquals(200, response.getStatusLine().getStatusCode());
+                assertEquals(asyncId, response.getHeader(HEADER_NAME_ASYNC_ID));
+                assertEquals("false", response.getHeader(HEADER_NAME_ASYNC_PARTIAL));
+                assertEquals("false", response.getHeader(HEADER_NAME_ASYNC_RUNNING));
+            }
+
+            cursor = response.getHeader(HEADER_NAME_CURSOR);
+            String body = responseBody(response);
+            if (i == fetchCount) {
+                assertNull(cursor);
+                assertFalse(hasText(body));
+            } else {
+                String expected = expectedTextBody(format, fetchSize, i, csvDelimiter);
+                assertEquals(expected, body);
+
+                if (hasText(cursor) == false) { // depending on index and fetch size, the last page might or not have a cursor
+                    assertEquals(i, fetchCount - 1);
+                    i++; // end the loop after deleting the async resources
+                }
+            }
+
+            // delete the query results
+            Request request = new Request("DELETE", SQL_ASYNC_DELETE_REST_ENDPOINT + asyncId);
+            Map<String, Object> deleteStatus = toMap(client().performRequest(request), null);
+            assertEquals(200, response.getStatusLine().getStatusCode());
+            assertTrue((boolean) deleteStatus.get("acknowledged"));
+        }
+    }
+
+    static Map<String, Object> runSql(RequestObjectBuilder builder, String mode) throws IOException {
+        return toMap(runSql(builder.mode(mode)), mode);
+    }
+
+    static Response runSql(RequestObjectBuilder builder) throws IOException {
+        return runSqlAsTextFormat(builder, null);
+    }
+
+    static Response runSqlAsTextFormat(RequestObjectBuilder builder, @Nullable String format) throws IOException {
```

Comment:
maybe a better name is `runSqlWithFormat`

Reply:
Makes sense, will change.

Reply:
Changed. to `runSqlAsTextWithFormat` since there are existing `runSqlAsText` methods and only text formats need to be provided (as it checks for a `Cursor` header, only available with those formats).